### PR TITLE
Allow adapter-only first-party harness use (10.1.1)

### PR DESCRIPTION
**Pull request by leynos**
## Summary

This branch implements roadmap task (10.1.1) by allowing first-party Tokio and GPUI adapter users to depend on the selected adapter crate without also listing `rstest-bdd-harness` directly. Custom harness authors and callers that import base harness API types still depend on `rstest-bdd-harness` explicitly.

Roadmap task: (10.1.1)
Execplan: [docs/execplans/10-1-1-first-party-adapters-compile-without-direct-base-harness-dependency.md](https://github.com/leynos/rstest-bdd/blob/971301bc291c79d7a02d7dffeb1fe4c1fed21aee/docs/execplans/10-1-1-first-party-adapters-compile-without-direct-base-harness-dependency.md)

## Review walkthrough

- Start with [crates/rstest-bdd-macros/src/codegen/mod.rs](https://github.com/leynos/rstest-bdd/blob/971301bc291c79d7a02d7dffeb1fe4c1fed21aee/crates/rstest-bdd-macros/src/codegen/mod.rs) and [crates/rstest-bdd-macros/src/codegen/scenario.rs](https://github.com/leynos/rstest-bdd/blob/971301bc291c79d7a02d7dffeb1fe4c1fed21aee/crates/rstest-bdd-macros/src/codegen/scenario.rs) to see how generated harness and attribute-policy assertions now resolve base API paths through canonical first-party adapter crate roots.
- Then review [crates/rstest-bdd-harness-tokio/src/lib.rs](https://github.com/leynos/rstest-bdd/blob/971301bc291c79d7a02d7dffeb1fe4c1fed21aee/crates/rstest-bdd-harness-tokio/src/lib.rs) and [crates/rstest-bdd-harness-gpui/src/lib.rs](https://github.com/leynos/rstest-bdd/blob/971301bc291c79d7a02d7dffeb1fe4c1fed21aee/crates/rstest-bdd-harness-gpui/src/lib.rs) for the adapter re-exports that generated code uses.
- Check [examples/tokio-reminders/Cargo.toml](https://github.com/leynos/rstest-bdd/blob/971301bc291c79d7a02d7dffeb1fe4c1fed21aee/examples/tokio-reminders/Cargo.toml) and [examples/gpui-counter/Cargo.toml](https://github.com/leynos/rstest-bdd/blob/971301bc291c79d7a02d7dffeb1fe4c1fed21aee/examples/gpui-counter/Cargo.toml) for the downstream-style adapter-only manifest proof.
- Finish with [docs/v0-6-0-migration-guide.md](https://github.com/leynos/rstest-bdd/blob/971301bc291c79d7a02d7dffeb1fe4c1fed21aee/docs/v0-6-0-migration-guide.md), [docs/users-guide.md](https://github.com/leynos/rstest-bdd/blob/971301bc291c79d7a02d7dffeb1fe4c1fed21aee/docs/users-guide.md), and [docs/rstest-bdd-design.md](https://github.com/leynos/rstest-bdd/blob/971301bc291c79d7a02d7dffeb1fe4c1fed21aee/docs/rstest-bdd-design.md) for the documented dependency matrix and design decision.

## Validation

- `cargo check -p tokio-reminders --tests`: passed after reproducing the pre-fix missing `rstest-bdd-harness` failure.
- `cargo check -p gpui-counter --tests`: passed.
- `cargo test -p rstest-bdd-macros`: passed.
- `cargo test -p tokio-reminders`: passed.
- `cargo test -p gpui-counter`: passed.
- `cargo test -p rstest-bdd-harness-gpui --features native-gpui-tests --test macro_compile -- --nocapture`: passed.
- `make check-fmt`: passed.
- `make lint`: passed.
- `make test`: passed, including 1389 nextest tests and 62 Python publish-check tests.
- `make markdownlint`: passed.
- `make nixie`: passed.

## Notes

The GPUI adapter-only compile proof exposed a `copy_dir_tree` staging edge case for nested missing destination parents. This branch fixes that helper with a focused regression test so the GPUI trybuild fixtures can stage their feature tree before compile validation.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[crates/rstest-bdd-harness/src/trybuild_staging/tests.rs](https://github.com/leynos/rstest-bdd/pull/487/files/971301bc291c79d7a02d7dffeb1fe4c1fed21aee#diff-02ff62dce2e9c0676949dfd4983db1582989ae98e2b97ef496ecc74132f3fb43)

Comment on file

~~~rust
    assert!(!dst.join("stale").exists());
}

#[rstest]
~~~

❌ Getting worse: [Code Duplication](https://codescene.io/projects/69391/delta?repo-id=1163913&review-id=487&biomarker=Code+Duplication&filename=crates%2Frstest-bdd-harness%2Fsrc%2Ftrybuild_staging%2Ftests.rs&method=copy_dir_tree_creates_missing_destination_parents)
introduced similar code in: copy_dir_tree_creates_missing_destination_parents

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[crates/rstest-bdd-macros/src/codegen/mod.rs](https://github.com/leynos/rstest-bdd/pull/487/files/971301bc291c79d7a02d7dffeb1fe4c1fed21aee#diff-d86875ec074a172f42b9142ecbd6e091970126470f971aab8a5532fb2288d739)

Comment on file

~~~rust
pub(crate) mod scenario;
pub(crate) mod wrapper;

const TOKIO_HARNESS_PACKAGE: &str = "rstest-bdd-harness-tokio";
~~~

❌ New issue: [String Heavy Function Arguments](https://codescene.io/projects/69391/delta?repo-id=1163913&review-id=487&biomarker=String+Heavy+Function+Arguments&filename=crates%2Frstest-bdd-macros%2Fsrc%2Fcodegen%2Fmod.rs&method=)
In this module, 63.2% of all arguments to its 12 functions are strings. The threshold for string arguments is 39.0%

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (3 warnings, 1 inconclusive)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| User-Facing Documentation | ⚠️ Warning | Two review-flagged documentation issues remain unresolved: second-person pronouns in users-guide.md violate style guidelines; canonical-path caveat missing from migration guide. | Replace second-person pronouns in users-guide.md (lines 871-874, 966-969). Add canonical-path requirement caveat to v0-6-0-migration-guide.md explaining that renamed/imported types require direct rstest-bdd-harness dependency. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Tests for `rstest_bdd_harness_api_path_for` lack coverage for renamed dependencies and aliased imports. Only canonical paths are tested; no validation of renamed/aliased fallback cases. | Add tests for renamed dependency resolution and aliased imports. Verify non-canonical paths correctly fall back to base harness. Explicitly document the canonical-path requirement in migration guide. |
| Testing (Property / Proof) | ⚠️ Warning | PR introduces path-resolution invariant across multiple input states but lacks property tests or model checking. Documentation omits canonical-path constraint. Review comments unaddressed. | Add parameterised tests for path matching. Caveat migration guide: adapter-only usage requires canonical crate-root paths. Fix pronouns in users-guide; collapse fallback tests using #\[rstest\]. |
| Testing (Compile-Time / Ui) | ❓ Inconclusive | No result was produced after verification. Marking as INCONCLUSIVE. | Re-run the check or adjust instructions to produce a final result. |

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

## Code Duplication
crates/rstest-bdd-macros/src/codegen/scenario/test_attrs.rs: 

### What lead to degradation?
The module contains 2 functions with similar structure: resolve_imported_harness_hint,resolve_imported_policy_hint

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

File: crates/rstest-bdd-macros/src/codegen/mod.rs [hunk 1]
```diff
@@ -28,68 +63,146 @@ pub(crate) fn rstest_bdd_harness_path() -> TokenStream2 {
...
-                FoundCrate::Itself => Ident::new(default_ident, Span::call_site()),
-                FoundCrate::Name(name) => Ident::new(&name, Span::call_site()),
-            };
-            quote! { ::#ident }
-        }
-        Err(err) => handle_missing_crate(crate_name_str, &err),
+/// Return the crate root that provides base harness API for the given harness
+/// or attribute-policy path.
+pub(crate) fn rstest_bdd_harness_api_path_for(adapter_path: &syn::Path) -> TokenStream2 {
+    first_party_adapter_api_path(adapter_path).unwrap_or_else(rstest_bdd_harness_path)
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-macros/src/codegen/mod.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (2 errors, 2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Tests verify trait names but not crate paths. Existing tests would pass even if generate\_trait\_assertions used incorrect crate paths. | Add tests asserting correct crate paths appear: ::rstest\_bdd\_harness\_tokio for TokioHarness, ::rstest\_bdd\_harness\_gpui for GpuiAttributePolicy, base harness for custom types. |
| Module-Level Documentation | ❌ Error | Two modules have insufficient module documentation. `scenario.rs` and `test_attrs.rs` contain only single-line docstrings lacking explanation of utility and component relationships. | Expand docstrings to explain utility and relationships. `scenario.rs` should document coordination of domain, helpers, metadata, runtime, and test\_attrs. `test_attrs.rs` should explain hint resolution and attribute token generation. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Unit tests for codegen cover edge cases well. Integration tests and publish validator import from base harness directly, not adapters. No test validates adapter re-exports suffice for consuming code. | Update publish\_check\_gpui\_manifest validator to import from rstest\_bdd\_harness\_gpui. Add behavioural test for adapter re-export paths. Add CI gate to verify examples compile without direct harness dependency. |
| Observability | ⚠️ Warning | PR introduces silent fallback from first-party adapter detection to base harness path when adapters renamed/aliased. No observability provided for this fallback decision point. | Add compile-time diagnostic when adapter detection fails. Emit proc\_macro warning/note indicating fallback occurred and advising users to use canonical paths or add rstest-bdd-harness directly. |

**Comment by leynos** [comment]
@coderabbitai please create a new GitHub issue for the following:

Emitting a compile-time warning from a stable proc macro requires either the nightly `proc_macro::Diagnostic` API or the `proc_macro_error2` crate. Adding that dependency, deciding on message wording, and threading the diagnostic through `rstest_bdd_harness_api_path_for` is a self-contained piece of work that extends beyond the scope of this PR.

The issue would cover:

- Adding `proc_macro_error2` (or awaiting stabilisation of `proc_macro::Diagnostic`) as a dependency of `rstest-bdd-macros`.
- Emitting a `warning` from `first_party_adapter_api_path` when the fallback path is taken, advising the user to use canonical crate-root paths or add `rstest-bdd-harness` directly.
- Adding a trybuild compile-output test to assert the warning text.

Seek to retain stable support for the core library and macros.

**Comment by leynos** [comment]
@coderabbitai pause

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[crates/rstest-bdd-macros/src/codegen/scenario/tests.rs](https://github.com/leynos/rstest-bdd/pull/487/files/f1d365ab5596c9c5a90fe62409abc013c1925038#diff-aadfa2732c75e4fd903f5c6030c35d108e22e4208040c952edd97b4bb57353a2)

Comment on file

~~~rust
    assert_eq!(scenario_allows_skip(&tags), expected);
}

// -----------------------------------------------------------------------------
~~~

❌ New issue: [String Heavy Function Arguments](https://codescene.io/projects/69391/delta?repo-id=1163913&review-id=487&biomarker=String+Heavy+Function+Arguments&filename=crates%2Frstest-bdd-macros%2Fsrc%2Fcodegen%2Fscenario%2Ftests.rs&method=)
In this module, 71.4% of all arguments to its 12 functions are strings. The threshold for string arguments is 39.0%

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational.

### ❌ Failed checks (2 errors, 2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Tests verify trait names but not crate paths. Existing tests would pass even if generate\_trait\_assertions used incorrect crate paths. | Add tests asserting correct crate paths appear: ::rstest\_bdd\_harness\_tokio for TokioHarness, ::rstest\_bdd\_harness\_gpui for GpuiAttributePolicy, base harness for custom types. |
| Module-Level Documentation | ❌ Error | Two modules have insufficient module documentation. `scenario.rs` and `test_attrs.rs` contain only single-line docstrings lacking explanation of utility and component relationships. | Expand docstrings to explain utility and relationships. `scenario.rs` should document coordination of domain, helpers, metadata, runtime, and test\_attrs. `test_attrs.rs` should explain hint resolution and attribute token generation. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Unit tests for codegen cover edge cases well. Integration tests and publish validator import from base harness directly, not adapters. No test validates adapter re-exports suffice for consuming code. | Update publish\_check\_gpui\_manifest validator to import from rstest\_bdd\_harness\_gpui. Add behavioural test for adapter re-export paths. Add CI gate to verify examples compile without direct harness dependency. |
| Observability | ⚠️ Warning | PR introduces silent fallback from first-party adapter detection to base harness path when adapters renamed/aliased. No observability provided for this fallback decision point. | Add compile-time diagnostic when adapter detection fails. Emit proc\_macro warning/note indicating fallback occurred and advising users to use canonical paths or add rstest-bdd-harness directly. |

**Comment by leynos** [comment]
@sourcery-ai review

**Comment by SourceryAI** [comment]
Sorry @leynos, you have reached your weekly rate limit of 2500000 diff characters.

Please try again later or [upgrade](https://app.sourcery.ai/login?connection=github&git_namespace_id=746767&utm_source=github&utm_medium=bot&utm_campaign=rate_limited) to continue using Sourcery
